src/h2_stream_set.erl

```diff
@@ -0,0 +1,855 @@
+    OldStream = lists:keyfind(Id, 2, PeerSubset#peer_subset.active),
+    OldType = type(OldStream),
+    ActiveDiff =
+        case OldType of
+            closed -> 0;
+            active -> -1
+        end,
+
+    NewActive = lists:keyreplace(Id, 2, PeerSubset#peer_subset.active, Closed),
+    PeerSubset#peer_subset{
+      active_count=PeerSubset#peer_subset.active_count+ActiveDiff,
+      active=NewActive
+     };
+%% Case 3: It's closed, but greater than or equal to next available:
+upsert_peer_subset(
+  #closed_stream{
+     id=Id
+    } = Closed,
+  PeerSubset)
+ when Id >= PeerSubset#peer_subset.next_available_stream_id ->
+    PeerSubset#peer_subset{
+      next_available_stream_id=Id+2,
+      active=lists:keystore(Id, 2, PeerSubset#peer_subset.active, Closed)
+     };
+%% Case 4: It's active, and in the range we're working with
+upsert_peer_subset(
+  #active_stream{
+     id=Id
+    }=Stream,
+  PeerSubset)
+  when Id >= PeerSubset#peer_subset.lowest_stream_id,
+       Id < PeerSubset#peer_subset.next_available_stream_id ->
+    PeerSubset#peer_subset{
+      active = lists:keystore(Id, 2, PeerSubset#peer_subset.active, Stream)
+     };
+%% Case 5: It's active, but it wasn't active before and activating it
+%% would exceed our concurrent stream limits
+upsert_peer_subset(
+  #active_stream{},
+  PeerSubset)
+  when PeerSubset#peer_subset.max_active =/= unlimited,
+       PeerSubset#peer_subset.active_count >= PeerSubset#peer_subset.max_active ->
+    {error, ?REFUSED_STREAM};
+%% Case 6: It's active, and greater than the range we're tracking
+upsert_peer_subset(
+  #active_stream{
+     id=Id
+    }=Stream,
+  PeerSubset)
+ when Id >= PeerSubset#peer_subset.next_available_stream_id ->
+    PeerSubset#peer_subset{
+      next_available_stream_id=Id+2,
+      active_count=PeerSubset#peer_subset.active_count+1,
+      active = lists:keystore(Id, 2, PeerSubset#peer_subset.active, Stream)
+     };
+%% Catch All
+upsert_peer_subset(
+ Stream,
+ PeerSubset) ->
+    lager:error("Tried upserting ~p to ~p, but clause isn't covered",
+                [Stream, PeerSubset]),
+    PeerSubset.
+
+
+drop_unneeded_streams(Streams) ->
+    SortedStreams = lists:keysort(2, Streams),
+    lists:dropwhile(
+      fun(#closed_stream{
+             garbage=true
+            }) ->
+              true;
+         (_) ->
+              false
+      end,
+      SortedStreams).
+
+-spec close(
+        Stream :: stream(),
+        Response :: garbage | {hpack:headers(), iodata()},
+        Streams :: stream_set()
+                   ) ->
+                   { stream(), stream_set()}.
+close(Stream,
+      garbage,
+      StreamSet) ->
+    Closed = #closed_stream{
+                id = stream_id(Stream),
+                garbage=true
+               },
+    {Closed, upsert(Closed, StreamSet)};
+close(Closed=#closed_stream{},
+      _Response,
+      Streams) ->
+    {Closed, Streams};
+close(_Idle=#idle_stream{id=StreamId},
```

Comment:
It wouldn't. the close function needs the ability to create a #closed_stream{} record from any input. most cases, this should be an #active_stream coming in, but in the case that an #idle_stream comes in, we don't want the dreaded `badmatch`